src/test/passes/string-encoding/basic-aarch64-stackopt.ll

```diff
@@ -0,0 +1,40 @@
+;
+; This file is distributed under the Apache License v2.0. See LICENSE for details.
+;
+
+; REQUIRES: aarch64-registered-target
+
+;     RUN: env OMVLL_CONFIG=%S/config_replace.py clang++ -fpass-plugin=%libOMVLL \
+;     RUN:         -target arm64-apple-ios17.5.0 -S -emit-llvm -O0 -c %s -o - | FileCheck %s
+;
+;     RUN: env OMVLL_CONFIG=%S/config_replace.py clang++ -fpass-plugin=%libOMVLL \
+;     RUN:         -target aarch64-linux-android -S -emit-llvm -O0 -c %s -o - | FileCheck %s
+;
+;     CHECK-NOT:     {{.*Hello, Stack.*}}
+
+@__const.main.Hello = private constant [13 x i8] c"Hello, Stack\00", align 1
+
+define i32 @main() {
+; CHECK-LABEL: @main(
+; CHECK:         %5 = getelementptr inbounds [13 x i8], ptr @0, i64 0, i64 0
+; CHECK-NEXT:    %6 = load i1, ptr @1, align 1
+; CHECK-NEXT:    %7 = icmp eq i1 %6, false
+; CHECK-NEXT:    br i1 %7, label %8, label %__omvll_decode_wrap.exit
+; CHECK:       8:
+; CHECK-NEXT:    call void @__omvll_decode(ptr %5, ptr @__const.main.Hello, i64 %3, i32 %4)
+; CHECK-NEXT:    store i1 true, ptr @1, align 1
+; CHECK-NEXT:    br label %__omvll_decode_wrap.exit
+; CHECK:       __omvll_decode_wrap.exit:
+; CHECK-NEXT:    call void @llvm.memcpy.p0.p0.i64(ptr align 1 %buffer, ptr align 1 %5, i64 13, i1 false)
+  %buffer = alloca [13 x i8], align 1
+  call void @llvm.lifetime.start.p0(i64 13, ptr %buffer)
+  call void @llvm.memcpy.p0.p0.i64(ptr align 1 %buffer, ptr align 1 @__const.main.Hello, i64 13, i1 false)
+  %puts = call i32 @puts(ptr %buffer)
```

Comment:
Is this test representative? In particular, that there is a buffer existing inline in the original IR?

I had an example like `void test() { puts("Hello, Stack"); }` in mind, that results in IR like this:
```
@.str = private unnamed_addr constant [13 x i8] c"Hello, Stack\00", align 1

define void @test() #0 {
  %1 = call i32 @puts(ptr noundef @.str)
  ret void
}
```

Would the pass apply here as well?

Reply:
Both tests are fine, the one I added uses an array buffer, yours uses a string literal.

Reply:
Well, the difference is that the string literal is not a writable location. But ok, I see now that test is just exercising a more complex case. The string literal is the (read-only) input and the array buffer is basically unrelated: The alloca could be optimized away and `%5` passed to `@puts` directly.

Reply:
The string literal is not in a writeable location _at runtime_, the purpose of the pass is exactly to encode constant strings. What I meant above, is that both examples use a constant global variable, which is what we are testing here. The one I added simply involves an extra array buffer instead of directly passing the literal to puts. I agree it's slightly more complex, so I'm removing it and using directly the literal.